source/sas_compress.cpp

```diff
@@ -280,24 +283,46 @@ std::string LZ4Compressor::compress(const std::string& s, const SAS::Profile* pr
       _saved_streams[profile] = saved_lz4_stream(base_stream, stream_saved_buf);
       saved_stream_iterator = _saved_streams.find(profile);
     }
-    
-    LZ4_stream_restore_preserved(_stream,
-                                 saved_stream_iterator->second.first,
-                                 saved_stream_iterator->second.second);
   }
 
+
+
   // Spin round, compressing up to a buffer's worth of input and appending it to the string.
   std::string compressed;
-  for (unsigned int s_pos = 0; s_pos < s.length(); s_pos += MAX_INPUT_SIZE)
+  bool success = false;
+  while (!success)
   {
-    // Because we've allocated a big enough buffer, it's not possible for this call to fail.
-    int buffer_len = LZ4_compress_fast_continue(_stream,
-                                                &s.c_str()[s_pos],
-                                                _buffer,
-                                                (s.length() - s_pos > MAX_INPUT_SIZE) ? MAX_INPUT_SIZE : s.length() - s_pos,
-                                                sizeof(_buffer),
-                                                ACCELERATION);
-    compressed += std::string(_buffer, buffer_len);
+    // Clear the stream after the last compression attempt.
+    LZ4_resetStream(_stream);
+
+    if (profile)
+    {
+      LZ4_stream_restore_preserved(_stream,
+                                   saved_stream_iterator->second.first,
+                                   saved_stream_iterator->second.second);
+    }
+
+
+    // Attempt to compress this data using the current buffer.
+    int compressed_len = LZ4_compress_fast_continue(_stream,
+                                                    s.c_str(),
+                                                    _buffer,
+                                                    s.length(),
+                                                    _buffer_len,
+                                                    ACCELERATION);
+
+    if (compressed_len <= 0)
+    {
+      // Compression failed - retry with a bigger buffer. We permanently
+      // enlarge this buffer so we aren't redoing this on every compression.
```

Comment:
I don't understand this comment, and I'm also concerned that we might have failed for reasons other than _buffer_len being too small - in which case we'd loop forever, right?